### PR TITLE
feat: Track stalls in the JavaScript event loop as measurements

### PR DESCRIPTION
**Pull request by jennmueng**
## :loudspeaker: Type of change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring

### Important TL;DR

- Tracks stalls in the JavaScript event loop as measurements attached to transactions.
- Is **enabled by default** as a default integration. Can be disabled by disabling the default integration.
- Will actually work on other Javascript platforms other than React Native. We should consider expending it if it turns out to work well.
- Monkey patches the `hub.startTransaction` and `transaction.finish` along with routing idleTransactions through `ReactNativeTracing` to instrument every transaction.
- Supports the `trimEnd` functionality where the end timestamp of the transaction is set to the last finished span.
- Does not support `startTimetamp` and `endTimestamp` being custom set.
- If you want to see it in action running on JS, try logging `measurements` in the if statement on line 32 here when you run the jest tests: https://github.com/getsentry/sentry-react-native/blob/0d573be96eb6865db6fcaaf4c0a4554d54d29247/test/integrations/stalltracking.test.ts#L6-L40

### How it works

Measures the stall time in the JS event loop by setting a timeout and seeing how long it strays from the expected time that it's supposed to run. By default, the "`minimumStallThreshold `" is set to `50ms`, with each timeout running every  `50ms` which allows a margin of error of `minimumStallThreshold  = 50ms` before the straying is considered a stall. The stall time is then calculated by taking the total difference from the time the timeout is supposed to be run — `totalTimeTaken - 50ms`.

This is inspired by the `JSEventLoopWatchdog` [used internally in React Native]( https://github.com/facebook/react-native/blob/006f5afe120c290a37cf6ff896748fbc062bf7ed/Libraries/Interaction/JSEventLoopWatchdog.js).

#### Spans

We monkey patch the finish method of spans that have been added to the `spanRecorder`. Every time a span inside a transaction is finished, we log the last end timestamp in the chance that `trimEnd` is used and the transaction's end timestamp is set to the last span's end timestamp.

### Example Measurement
```json
{
	"measurements": {
	  "stall_count": { 
			"value": 2 
		},
	  "stall_total_time": { 
			"value": 247.875 
		}, 
	  "stall_longest_time": { 
			"value": 175.125 
		},
	}
}
```

### Why we don't use this method to measure slow and frozen JS frames:
We would need to set the timeout to 16.67ms (for 60fps) and log any iterations that take longer than that. However, in practice when doing so it ends up being incredibly flakey, with the majority of the event loop iterations taking longer than the 16.67ms needed for 60fps and it ends up logging almost all frames as slow. This is not accurate to what the fps performance monitor is logging when using RN in dev mode.

## :bulb: Motivation and Context

Tracking JS event loop stalls should be correlated to JS Frame drops. Although we would prefer to track frame drops to align the behavior with the native SDKs, this is the only way to track a measurement that represents app slowdowns on the JS layer that I can find.

This method is inspired by how [React Native measures stalls internally.]( https://github.com/facebook/react-native/blob/b176617d1e0cb7834698f147db4522290098ea00/Libraries/Interaction/JSEventLoopWatchdog.js)

## :green_heart: How did you test it?
**Extensive** unit tests, tested on Android and iOS sample app.

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [x] All tests passing
- [x] No breaking changes

## Next Steps

We also discussed logging each stall as a child span, which I believe would be massively helpful in pinpointing exactly when slowdowns happened in your app.

### CODE REVIEW

File: src/js/measurements.ts
Comment (jennmueng) [inline src/js/measurements.ts:66]: Measures the stall duration via setting a timeout to run, in this case, in 40ms, and then measures if the actual "busy time" is greater than 80ms. These numbers are just hard coded and arbitrary for now.

In theory we could measure slow and frozen frames using this by running the timeout every 16.667ms (60FPS / 1000ms) given we determine what duration constitutes "slow" frames and "frozen" frames. We should also note that running the timer every 16.667ms could have an effect on the performance of the app itself.

File: src/js/measurements.ts
Comment (jennmueng) [inline src/js/measurements.ts:125]: If a transaction has a custom set `endTimestamp` we do not add the stall measurements. Should do the same with and check if `startTimestamp` was custom set.

File: src/js/measurements.ts
Comment (jennmueng) [inline src/js/measurements.ts:88]: Here, we patch `hub.startTransaction()`, which is also aliased as `Sentry.startTransaction()`.

File: src/js/measurements.ts
Comment (bruno-garcia) [inline src/js/measurements.ts:129]: If we get multiple concurrent transactions each one will have its own tracker, right>
So multiple jobs being queued for execution to measure latency?

Comment (jennmueng) [reply src/js/measurements.ts:129]: Hmm yes, I realized that can have an effect on performance. If we decide to have one global one, it should not be always tracking and only when some transaction is running, and is closed when all the transactions are finished. Should I go with that?

**Comment by jennmueng** [comment]
As discussed on the grooming, this will be updated to send

```json
{
	"measurements": {
	  "stall_count": { 
			"value": 2 
		},
	  "stall_total_time": { 
			"value": 247.875 
		}, 
	  "stall_longest_time": { 
			"value": 175.125 
		},
	}
}

```

File: src/js/measurements.ts
Comment (bruno-garcia) [reply src/js/measurements.ts:129]: It would be awesome to have this. If it doesn't degrade performance in any obvious way it sounds good. It must have an opt-out though so folks can turn this off if it becomes an issue

Comment (jennmueng) [reply src/js/measurements.ts]: Added this feature.

**Comment by jennmueng** [comment]
Requested review from @getsentry/team-webplatform due to how heavy JS this PR is along with being applicable to the rest of the JS universe.

File: src/js/integrations/stalltracking.ts
Comment (AbhiPrasad) [inline src/js/integrations/stalltracking.ts:1]: This is a nice approach - will do a more detailed review in a bit. What do you think we need to do to move this into sentry-javascript.

Also the monkey patching is not ideal, we should update tracing to make it easier to patch rather than patch the idle transaction this way tbh.

Comment (jennmueng) [reply src/js/integrations/stalltracking.ts:1]: Yeah I strongly agree the monkey patching is bad, but we reckoned that we're going to release this as a prerelease to see how it goes, if it's actually effective, and if it's valuable to customers, if it turns out well we can then move it over to the javascript repo and solidify it by updating tracing to actually support hooks to the start and finish methods.

Comment (AbhiPrasad) [reply src/js/integrations/stalltracking.ts:1]: Sounds like a plan 👍

File: src/js/integrations/stalltracking.ts
Comment (AbhiPrasad) [inline src/js/integrations/stalltracking.ts:90]:
```suggestion
      // noop
      return () => null;
```

File: src/js/integrations/stalltracking.ts
Comment (AbhiPrasad) [inline src/js/integrations/stalltracking.ts:217]: Nice, forgot that maps keep insertion order

File: src/js/integrations/stalltracking.ts
Comment (AbhiPrasad) [inline src/js/integrations/stalltracking.ts:240]: Can we use an `array.reduce` here?

File: src/js/integrations/stalltracking.ts
Comment (AbhiPrasad) [inline src/js/integrations/stalltracking.ts:326]: Should we have a `_isTracking` check here?

File: test/integrations/stalltracking.test.ts
Comment (AbhiPrasad) [inline test/integrations/stalltracking.test.ts:20]:
```suggestion
    const expensiveObject: { value: string[] } = { value: Array(100000).fill("expensive") };
```

File: test/integrations/stalltracking.test.ts
Comment (AbhiPrasad) [inline test/integrations/stalltracking.test.ts:56]: Can we use jest's `done()` callback here?

**Review by AbhiPrasad — CHANGES_REQUESTED** [review]
Give me some more time to think about this implementation, but on a high level I like the direction.

File: src/js/integrations/stalltracking.ts
Comment (jennmueng) [reply src/js/integrations/stalltracking.ts:19]: From the user point of view don't we support users passing either ms or sec or did I misunderstand?

File: src/js/integrations/stalltracking.ts
Comment (jennmueng) [reply src/js/integrations/stalltracking.ts:217]: Haha yeah when I was about to write this and go through the keys I was like wait, they're already in insertion order lol

File: src/js/integrations/stalltracking.ts
Comment (AbhiPrasad) [reply src/js/integrations/stalltracking.ts:19]: Will defer to @rhcarvalho for the most correct answer here

File: package.json
Comment (jennmueng) [reply package.json:17]: Yeah and it's hanging the CI right now so it's not good. I'm in favor of removing it, but maybe we can just expose `_timeout` and `_isTracking` to jest test that the timeout iterations have stopped?

Comment (jennmueng) [reply package.json:17]: Or with an `_isTracking` check in `iteration` just check for `isTracking: false`

Comment (AbhiPrasad) [reply package.json:17]: I'm a +1 on `_isTracking` check

File: test/integrations/stalltracking.test.ts
Comment (jennmueng) [reply test/integrations/stalltracking.test.ts:52]: Oh yeah I somehow forgot all about `done()` 😅

File: src/js/index.ts
Comment (jennmueng) [reply src/js/index.ts:44]: Hmm I'm not too familiar with side effects (or tree shaking for that matter), is this only a web/webpack thing?

**Comment by jennmueng** [comment]
After some deliberating. I will remove the transaction `endTimestamp` support and reduce the span finish tracking to just the last one which will greatly reduce complexity.

File: src/js/index.ts
Comment (AbhiPrasad) [reply src/js/index.ts:44]: Yeah it's a webpack thing - no idea if it applies in React native land

Comment (jennmueng) [reply src/js/index.ts:44]: If that's the case, then we're good as React Native doesn't use webpack but rather their own bundler—Metro

**Comment by jennmueng** [comment]
- Removed the complex support for `endTimestamp` and instead we just support `trimEnd` being used. Although due to the different behavior regarding unfinished spans, we handle idleTransactions and normal transactions differently. Please double check the logic I used for this here:
https://github.com/getsentry/sentry-react-native/blob/9ddd3e9caaa7ab4d9ee668f4e4364104f23a58a5/src/js/integrations/stalltracking.ts#L149-L187

- Updated tests to test the latest implementation along with adding new tests:
  - Multiple stalls are counted for
  - Idletransaction with unfinished spans
  - Unfinished spans in idle transactions
  - Only measures stalls within the transaction duration when `trimEnd` is used

File: src/js/integrations/stalltracking.ts
Comment (AbhiPrasad) [reply src/js/integrations/stalltracking.ts:137]: Also means we can remove the ternary on line 159.

File: src/js/integrations/stalltracking.ts
Comment (AbhiPrasad) [inline src/js/integrations/stalltracking.ts:312]: Can we leave a comment why timeoutDuration is half of the acceptable busy time? I kind of understand why, but I'm sure future readers would appreciate the context.

File: src/js/integrations/stalltracking.ts
Comment (AbhiPrasad) [inline src/js/integrations/stalltracking.ts:302]: let's try to avoid creating new objects.

```suggestion
    this._longestStallsByTransaction.clear();
    this._runningTransactions.clear();
    this._statsAtTimestamp.clear();
```

File: src/js/integrations/stalltracking.ts
Comment (AbhiPrasad) [inline src/js/integrations/stalltracking.ts:325]: nit: Can we extract this into a variable?

**Review by AbhiPrasad — COMMENTED** [review]
gimme another day to reflect on this, but tbh we can probably ship it

File: src/js/integrations/stalltracking.ts
Comment (jennmueng) [reply src/js/integrations/stalltracking.ts:226]: `_markSpanFinish` it is

File: src/js/integrations/stalltracking.ts
Comment (jennmueng) [reply src/js/integrations/stalltracking.ts:310]: Hmm the original code from Facebook actually used 1/5th the "acceptable busy time", the number is pretty arbitrary and up to how often we want to run the iterations and what we consider a stall.

I came up with a better solution is to have `minimumStallThreshold` instead of `acceptableBusyTime`. `timeoutDuration` would instead be a fixed `50ms` (or whatever we want to set it to, while keeping in mind the performance impact of having a lower number, and the lower detection sensitivity of the higher number), and a stall would now instead be logged on `totalTimeTaken  > 50ms + minimumStallThreshold`. How does that sound?

File: src/js/integrations/stalltracking.ts
Comment (AbhiPrasad) [inline src/js/integrations/stalltracking.ts:83]: nit: I would just remove, doesn't add much as a comment
```suggestion
```

File: src/js/integrations/stalltracking.ts
Comment (AbhiPrasad) [reply src/js/integrations/stalltracking.ts:312]: Yeah I like the `minimumStallThreshold` idea, plz include a comment in the code talking about how we switched from `acceptableBusyTime` -> `minimumStallThreshold` so we preserve this context.

File: src/js/sdk.ts
Comment (AbhiPrasad) [inline src/js/sdk.ts:83]: Should we only enable if tracing is active?

File: src/js/sdk.ts
Comment (AbhiPrasad) [inline src/js/sdk.ts:28]: One thought, do we want to make it not default enabled for this release, ask users to try it out manually, and then enable by default in the release after that?

**Review by AbhiPrasad — APPROVED** [review]
Spent the weekend thinking about it, and ok let's ship it. I would get a mobile team +1 just to confirm before you merge though. Also please make sure to address the default integration thing as well.

File: src/js/sdk.ts
Comment (jennmueng) [reply src/js/sdk.ts:28]: @bruno-garcia What do you say. Although this would be released in a beta, if we just enable it, it might work because the people who would be trying this out would be doing so just for it anyways? I'm fine with either.

**Comment by jennmueng** [comment]
Changed so tracing needs to be enabled for the `StallTracking` integration to be added along with tests for it.

File: src/js/integrations/stalltracking.ts
Comment (bruno-garcia) [inline src/js/integrations/stalltracking.ts:56]: Can we use `WeakMap`? We don't need to have a strong reference here and it could leak if for some reason `finish` isn't called.
This suggestion applies to any map of <Transaction, X>

File: src/js/integrations/stalltracking.ts
Comment (bruno-garcia) [inline src/js/integrations/stalltracking.ts:61]: We could use `_longestStallsByTransaction` for all operations below and get spare this extra Set

File: src/js/integrations/stalltracking.ts
Comment (bruno-garcia) [inline src/js/integrations/stalltracking.ts:158]: I think this is just the current state of affairs but we should have a chat with the team and align one way or the other.

Doesn't mean this blocks the PR, but it's something we need to address.

File: src/js/integrations/stalltracking.ts
Comment (jennmueng) [reply src/js/integrations/stalltracking.ts]: Replaced with a max running transaction count as discussed due to WeakMap not allowing us to iterate through the keys or with a way to know how many transactions we are tracking.

File: src/js/integrations/stalltracking.ts
Comment (AbhiPrasad) [inline src/js/integrations/stalltracking.ts:386]: Feel like this having this logic means that we should probably be using an LRU cache instead of a Map.

But I guess it's fine to keep it as it is. An explicit for loop will be more efficient than slice + forEach here.

```ts
let counter = 0;
const transactions = this._statsByTransaction.keys();
const len = this._statsByTransaction.size - MAX_RUNNING_TRANSACTIONS
for (let t of transactions) {
    if (counter > len) break;
    counter++;
    this._statsByTransaction.delete(t);
}
```

**Review by bruno-garcia — APPROVED** [review]
Can we see this flying?

**Review by AbhiPrasad — APPROVED** [review]

**Pull request merged**
